org-code-javabuilder/lib/src/main/java/org/code/neighborhood/GridSquare.java

```diff
@@ -0,0 +1,122 @@
+package org.code.neighborhood;
+
+public class GridSquare {
+  private String color;
+  private boolean passable;
+  private int paintCount;
+
+  private enum SquareType {
+    WALL,
+    OPEN,
+    START,
+    FINISH,
+    OBSTACLE,
+    STARTANDFINISH,
+    UNKNOWN
+  };
+
+  private SquareType squareType;
+
+  protected GridSquare(int tileType, int value) {
+    this.setTileType(tileType);
+    this.paintCount = value;
+    this.color = "";
+  }
+
+  protected GridSquare(int tileType) {
+    this.setTileType(tileType);
+    this.paintCount = 0;
+    this.color = "";
+  }
+
+  private void setTileType(int tileType) {
+    switch (tileType) {
+      case 0:
+        this.squareType = SquareType.WALL;
+        this.passable = false;
+        break;
+      case 1:
+        this.squareType = SquareType.OPEN;
+        this.passable = true;
+        break;
+      case 2:
+        this.squareType = SquareType.START;
+        this.passable = true;
+        break;
+      case 3:
+        this.squareType = SquareType.FINISH;
+        this.passable = true;
+        break;
+      case 4:
+        this.squareType = SquareType.OBSTACLE;
+        this.passable = false;
+        break;
+      case 5:
+        this.squareType = SquareType.STARTANDFINISH;
+        this.passable = true;
+        break;
+      default:
+        this.squareType = SquareType.UNKNOWN;
+        this.passable = false;
+    }
+  }
+
+  // Sets the color of the square to the given color
+  public void setColor(String color) {
+    if (!ColorHelpers.isColor(color)) {
+      System.out.println("Invalid color, please check your color format");
+      return;
+    }
+    if (this.passable && this.paintCount == 0) {
+      this.color = color;
+    }
+  }
+
+  // Determines whether the given coordinate can be moved into
+  public boolean isPassable() {
+    return this.passable;
+  }
+
+  // Decreases the paintCount by 1 if there is available paint
+  public void collectPaint() {
+    if (this.containsPaint()) {
+      this.paintCount--;
+    } else {
+      System.out.println("There's no paint to collect here");
```

Comment:
Check in with curriculum for these cases. We may want to set these as exceptions too.

Reply:
Unfortunately, this was a point of ambiguity on the original API. Figuring these out now will save us a lot of time in the future.

Reply:
We decided to just show the error message instead of stopping the program over being out of paint (the painter just wouldn't actually pick up any paint the way that Painter.takePaint is written) and we're following the same pattern for when the painter tries to paint without anything in their bucket.